doc/MiniLanguageTutorial.md

```diff
@@ -480,37 +563,278 @@ Mini never automatically converts types to make an operation succeed.  Programme
+> In the case of a file name followed by a identifier, the file name represents the mini file within the source folder with that name.
+> So if `"arb_os"` is the compile target `name` would represent `arb_os/name.mini`, and the identifier would represent the function or type of the same name in that file.
+> If it starts with a library prefix, the base folder is determined by the prefix, with
+> `std` corresponding to the `std` folder and `core` corresponding to the `builtin` folder.
+> The second and third parts of the path work the same as mentioned previously.
+> A path represented in this format will be called a virtual path, to distinguish it from paths in use by the filesystem.
+
+## Shadowing
+
+In mini, new variables can be created with the same name as old variables, this is called shadowing.
+To prevent ambiguity when dealing with shadowed variables, a reference to the shadowed name is treated as referring to
+the most recently declared variable that is currently in scope. So if a variable is declared inside a codeblock with the same name as another variable outside the codeblock,
+references to the name inside the codeblock will refer to the inner variable, and references to the name after the end of the codeblock refer to the outer variable.
+
+## AVM Values
+
+An AVM value may be one of the following:
+
+1. `Int`: a 256 bit integer
+1. `CodePoint`: a codepoint, points to an instruction in the code.
+1. `Buffer`: a buffer that contains bytes.
+1. `Tuple(`...`)`: up to 8 AVM values.
+
+More complex values are usually encoded by storing multiple values inside a `Tuple`, included nested tuples.
+
+## Default Value
+
+Every type has a default value that is used for the initial value of globals with that type.
+The default values for each type are as follows:
+
+* `uint`, `int`, `bytes32`, `address`:  
+  `0`, represented in `AVM` as an `Int`.
+* `bool`:  
+  `false`, this is also a `0` `Int` in the underlying AVM.
+* `buffer`:  
+  A buffer with length `0`, attempting to read any bytes past the end of a buffer will be treated as reading `0` values.
+* `any`:  
+  A length `0` tuple, `()`.
+* `union`:  
+  The default value of a union type is the default value of its leftmost variant.
+* `tuple`:  
+  The default value of a tuple is a wrapping tuple, with leaves equal to the default values of each field type respectively. 
+* `struct`:  
+  The default value of a struct is a wrapping tuple, with leaves equal to the default values of each field type respectively.
+* `option`:  
+  The default value of all `option` types is the `None` variant, in AVM this is represented as a length 1 tuple containing an `Int` value `0`
+* `func`:  
+  The default value of all `func` types is a `CodePoint` 
+* `map`:  
+  An empty map, in AVM this is represented as a length 2 `Tuple` with both fields as an `Int` with value `0`.
+* `array`:  
+  An `array` of *type* has a default value of an array with length 1 containing the default value of *type*.
```

Comment:
Need to complete this one.   And isn't the representation described an array of length zero?

Reply:
I've updated this. The default value is indeed of length 1, please see: [ast.rs](https://github.com/OffchainLabs/arb-os/blob/documentation/src/compile/ast.rs#L832)
and also these code examples:
```
var illustrate: [](uint, int);

view func main() -> [](uint, int) {
    return illustrate;
}
```
which returns: `(1, 1, (1, 1, ((0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0))))`
and:
```
func main() -> [](uint, int) {
    return newarray<(uint, int)>(0);
}func main() -> [](uint, int) {
    return newarray<(uint, int)>(0);
}
```
which returns: `(0, 1, ((0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0)))`

Reply:
I realized that there is actually a bug here after posting this comment. I'll update the docs to reflect this.